[WIP] org-mode support

nico202 opened the pull request:
PR for #35

ericsagnes commented on derivation.nix:
```diff
@@ -1,4 +1,5 @@
-{ stdenv, caddy, asciidoctor
+{ stdenv, caddy, asciidoctor, withEmacs ? false
+, emacspkg ? null, which
```
Having orgmode support by default will be more user friendly than having users to set derivation variables.

I would prefer to have `emacsPkg` set to a basic emacs package that can compile most of standard orgmode content but that can be changed for more specific needs.
This also allow to remove `withEmacs`

What do you think?

nico202 replied:
I do agree! I wasn't sure that having emacs as a dependency was ok for everybody, so this was just-in-case

ericsagnes commented on derivation.nix:
```diff
@@ -27,7 +28,7 @@ stdenv.mkDerivation rec {
     perlPackages.ImageExifTool
     (python27.withPackages (ps: [ ps.parsimonious ]))
-  ];
+  ] ++ (if withEmacs then [ emacspkg ] else []);
```
Nit, if wont be needed if `withEmacs` is removed, but the idiomatic nix way for this kind of logic is to use `lib.optional`, eg: `lib.optional withEmacs [ emacspkg ]`

nico202 replied:
Yes sorry, I never wrote things in nix except for my configuration/shell.nix and some simple package. Thanks for taking the time to explain the nix-way, I'd like to learn it but finding the time is always difficult

ericsagnes commented on src/lib/data.nix:
```diff
@@ -6,6 +6,9 @@ with (import ./utils.nix lib);
 let
+  withEmacs = true; # How to propagate from config?
+  emacspkg = pkgs.emacsWithPackages
+      (epkgs: (with epkgs.melpaStablePackages; [ use-package ]));
```
It should not be required to propagate from the derivation; As emacs would be in styx's `propagatedBuildInputs`, it will be available in a build env with styx in `buildInputs`, see https://github.com/styx-static/styx/blob/master/src/lib/data.nix#L91 for example.

nico202 replied:
Right, removing

ericsagnes commented on src/lib/data.nix:
```diff
@@ -56,8 +65,11 @@ let
+        "${preprocess."${markupType}"} ${fileData.path} > preprocessed"
+  else "cp ${fileData.path} preprocessed"}
+        python ${pkgs.styx}/share/styx/tools/parser.py < preprocessed > $out
+  '';
```
This should not be needed, the running order is parsing -> processing.
Parsing take a markup file with styx specific markup and turn it in a nix attribute set, containing `pages`, `content` and so on, that are totally cleaned of styx specific markup;
These attributes are then fed to `markupToHtml` to convert it to html in https://github.com/styx-static/styx/blob/master/src/lib/data.nix#L62-L67.

So just setting a `command.orgmode` should be enough.

nico202 replied:
"Parsing take a markup file with styx specific markup and turn it in a nix attribute set, containing pages, content and so on"

The preprocessing is creating the styx specific markup out of the org standard tags. eg. in org you can use #+title: title to set the document title. This is compatible with the html exporter, with the latex exproter, beamer, and any other exporter, so by writing an org file you just assume it's the same. The preprocess takes care of creating the styx "title", "tags" and "draft" attributes.

Without the first step, one should manually write those elements in a non-org sintax (the styx one)

ericsagnes replied:
Ok, I understand better now the need for preprocessing.

> The preprocess takes care of creating the styx "title", "tags" and "draft" attributes.

Metadata is a totally free format, there are case where other attributes are used, eg author, date or any arbitratry attribute like [level or banner](https://github.com/styx-static/styx/blob/master/src/scaffold/sample-data/posts/2016-09-16-themes.md). Will the preprocessing approach allow to set any attribute?

ericsagnes replied:
Also is the preprocessing step compatible with [embedded nix](https://styx-static.github.io/styx-site/documentation/index.html#data.embedded-nix)? Having embedded nix is nice as it allows to do  in site links and template calls from the content. (I guess this would bring the problem of non compatible orgmode content again)

ericsagnes replied:
I just had a thought, but in case it is hard to find a good tradeoff, what about using a flexible approach:

- If there is no styx metadata attribute at the top of the file, generate it like done in the preprocessing
- If there is a styx metadata attribute at the top of the file, just use it and don't generate one

This could fit both use cases, what do you think?

nico202 replied:
For the compatibility with embedded nix: it works! (and I love the fact that it is. I'm not doing anything, but luckily  this does not conflict with the org sintax. in org you have {{{macros}}} with 3 )
example:
```org
\{{ toString (2 + 2) }} returns: {{ toString (2 + 2) }}
```
becomes:
```nix
{{ toString (2 + 2) }} returns: 4
```

Updated example, here again: https://blog.nixo.xyz/posts/2018-01-26-org-mode.html

nico202 replied:
About the metadata: if _any_ metadata available, the lowest friction I can imagine is:
export the default fields, that are title, date, author, email, language, creator, by default, and then have a custom keyword like:
```org
#+STYX: option:value
```

where one can set styx-specific ones. It seems to me the lowest friction/best default behavior. If a different behavior is needed, it's simply possible to replace the org-preprocess/org-compile functions in the `.el` file

```org
#+TITLE: 2018-01-26-org-mode
#+DATE: <2018-02-01 Thu>
#+AUTHOR: Nicolò Balzarotti
#+EMAIL: anothersms@gmail.com
#+LANGUAGE: en
#+CREATOR: Emacs 25.3.1 (Org mode 9.1.6)
```

nico202 replied:
Sorry I read the last comment  ("This could fit both use cases, what do you think?") only now. I'm not sure if this approach is better than the previous. letting #+STYX: option:value override params like title, author and so on is the best approach I can think of, so that the file continue to be export-able in any other org-supported format without compromises (ie, having a {{-- at the top)

nico202 replied:
I'll try to explain this better in the next commit

ericsagnes replied:
The `#+STYX: option:value` way is great!

ericsagnes commented on src/tools/emacs-compile:
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env bash
+
+emacs --file "$1" --script $(dirname "$0")/org-support.el -f compile-org-file
```
This should be moved to `commands.orgmode` and it might be better to dynamically generate `org-support.el` (see a comment bellow for details).

ericsagnes commented on src/tools/org-support.el:
```diff
@@ -0,0 +1,119 @@
+
+(defun jk-org-kwd (KEYWORD)
+  "get the value of a KEYWORD in the form of #+KEYWORD: value"
+  (cdr (assoc KEYWORD (jk-org-kwds))))
```
I am not really sure about what this is doing, but this seems to do hardcode some very specific things.
For example, I see that this automagically generate the styx metadata, it seems very useful, but is this possible for a user to add custom metadata attributes?
Also, users might want to have different language support.

I would prefer to have a more generic approach allowing users to customize to their needs the orgmode functionality.
Just an idea, but the styx configuration system could be used to provide options to configure orgmode, and this file could be generated on the fly according to these options (note that this is not possible for the moment as the configuration is only used for themes and not propagated to lib), what do you think?

nico202 replied:
This is simply an helper function that gets from the buffer (the file) the value assigned to a "keyword". This simply mean that if in a buffer there's:

```org

#+title: This is the title
``` 
by calling `(jk-org-kwd "TITLE")` it returns "This is the title". The title tag is standard in org-mode. I use the same function on "draft" since it's the tag used by ox-hugo (another org-mode web exporter) to get if the file is a draft or it isn't (#+draft: t -> draft).

I'll comment about the customization thing in the next commit

ericsagnes commented on default.nix:
```diff
@@ -1,3 +1,11 @@
+      org-plus-contrib
+      htmlize
+    ]));
+}
```
Prefer sane defaults in `derivation.nix`, see next comment for details.

nico202 replied:
I'm not really sure on how to do it. 
```nix
, emacsWithPackages
, emacsPkg ? (emacsWithPackages (epkgs:
    (with epkgs.melpaPackages; with epkgs.orgPackages;
      [ use-package org htmlize ])))
, callPackage }:
```

Like this? it seems ugly

ericsagnes commented on src/lib/data.nix:
```diff
@@ -15,18 +18,24 @@ let
     # markups
     asciidoc = [ "asciidoc" "adoc" ];
     markdown = [ "markdown" "mdown" "md" ];
+    org-mode = if withEmacs then [ "org" ] else [];
```
-> `orgmode = [ "org" ];`

nico202 commented on src/lib/data.nix:
```diff
@@ -11,6 +11,126 @@ let
     emacs -Q --script ${styx-support} --file "$2" -f $1
   '';
 
+  styx-support = pkgs.writeText "styx-support.el" ''
```
The customization is the most important part in this exporter, since one should be able to include other emacs packages and customize it with elisp. I'd clean up the following config to its bare minimum and then add an extraConfig param that get appended to the end of the file, AND a config param that totally replaces it.

nico202 replied:
Also, I'd move this and it's customization options to a separate nix file. Any naming suggestion?

nico202 commented on src/lib/data.nix:
```diff
@@ -32,11 +152,11 @@ let
     asciidoc = "asciidoctor -b xhtml5 -s -a showtitle -o-";
     markdown = "multimarkdown";
-    org-mode = "bash ${pkgs.styx}/share/styx/tools/emacs-compile";
+    org-mode = "${org-compile} compile-org-file";
```
I need a separate command since the order of the given params is important, so I can't just add --file ${fileData.path}  at the end of the command. The last thing must be -f "function-to-call".

ericsagnes replied:
Ok, this is just a detail that can be cleaned up later. Thanks for the explanation.

nico202 commented on src/lib/default.nix:
```diff
@@ -5,7 +5,7 @@ let
   # nixpkgs lib
   base = nixpkgs.lib // builtins;
-  pkgs = { inherit styx; inherit (nixpkgs) runCommand writeText; };
+  pkgs = { inherit styx; inherit (nixpkgs) runCommand writeText writeScript bash; };
```
See the previous comment on why I need writeScript. bash is for the script shebang

ericsagnes commented:
There is a test mechanism for markup file loading and processing in https://github.com/styx-static/styx/blob/master/tests/lib.nix#L114.
It could be used to automatically check that orgmode files are correctly processed.

nico202 commented:
Yes! I'll implement the tests as soon we are features-stable ;)

nico202 commented:
Can you hint me on how to add an emacsExtraConf (string appended to the `.el` file) and a emacsConf (file that replaces the `.el` file entirely) options?
edit: or something similar

ericsagnes commented:
> Can you hint me on how to add an emacsExtraConf (string appended to the .el file) and a emacsConf (file that replaces the .el file entirely) options?

I need to think a little more about this one.

nico202 commented:
OMG I had not understood the power of styx until I read test `complex.md`.

I changed then a bit to allow any nix expression as a #+styx option (I was string-escaping them). Now this is really powerful I think. I'll write the missing tests and push the commits in the following hours. Then I'll wait for your idea on how to manage the extra config, that is not *that* important since it's already possible to define elisp directly inside the org file. So it's a nice-to-have low-prio thing imho

nico202 commented:
A bit explanation on the commit i'll push: its just an elisp rewrite/refactor in order to allow the following behaviour:

If a default org-property is defined, like the `#+title:` or the `#+author:`, use them. They are automatically converted into raw strings, in order to be compatible with other export engines.
To override those OR to create new ones, you can use: `#+styx-title:`, or `#+styx-anything:`. In this case, the value is kept as it is (so for the title, one needs to manually escape it like

```org
#+title: no quotes on the org property (this property will be overritten by styx-title)
#+author: author is kept "as it is" because no styx-author is defined
#+styx-title: "quoted styx-property because it's nix indeed"
#+styx-variable: { x = 8 + 1; }
#+styx-tags: [ "you" "can" "even" "do" "this" ]
#+styx-titleauhtor: "${title} - ${author}"
 ```

nico202 commented:
Here we are. Things not working:
1. Multiple pages (I mean, the do work but the breaks all much whole-file org-mode operations. No specific syntax in org right now because of this, but <<< works)
2. mixed-case meta.variables. that's because #+variables are capitalized automatically. I'm lowercasing them, but there's no an easy way to keep the capitalization I can think of right now.

However I'm kinda satisfied. Let me know what do you think. As usual I updated the rendering on the usual link, here I attached the source file. But tests should be quite explanatory
[source.txt](https://github.com/styx-static/styx/files/1687418/source.txt)

ericsagnes commented:
Thanks for all the hard work!

I have thought of a satisfying way to make the used package to compile markups configurable. It will be using a nixpkgs / styx-themes like approach.

This implies to change how styx is working but should result in a lighter and simpler software.
This will take some time and introduce some breaking changes.

So I am looking to make a Styx 1.0.0 release with these changes and the orgmode.

ericsagnes commented:
> Multiple pages (I mean, the do work but the breaks all much whole-file org-mode operations. No specific syntax in org right now because of this, but <<< works)

Right, I am thinking to change styx specific tags to markup language native tags. Does orgmode have a native page break tag that could be used in place of `<<<`.

For markup languages that does not have native page breaks like markdown, I am thinking to use 2 horizontal rules as a page break sign.

> mixed-case meta.variables. that's because #+variables are capitalized automatically. I'm lowercasing them, but there's no an easy way to keep the capitalization I can think of right now.

This is fine with me, we can just ignore this and add a note in the docs explaining that metadata keys should be lowercase.

nico202 commented:
> Does orgmode have a native page break

Not that I'm aware of, indeed.

Using a custom one should however be possible by customizing the export code, I might think of it in the next week, but probably I should ask some guru on emacs.stackexchange.com

kaushalmodi commented:
Hello, I don't use Styx, but by the magic of Internet, I just landed up here. How about using the latex syntax `\pagebreak`? Org doesn't support that pagebreak syntax for all exporters by default, but I have seen latex syntax borrowed by Org in many places.

nico202 commented:
@kaushalmodi  ahah yes the thing that you landed here is strange enough. Thanks for the suggestion, the problem is that the html exporter exports a single html page even in presence of `\pagebreak`. It can be used as an alterantive syntax for the current styx `<<<` but it's not substantially different. 

I agree that it's better than the arbitrary `<<<`, but having a `#+styx: split` in the same way [`org-reveal`](https://github.com/yjwen/org-reveal/) does, but adding support for multipage html export is something to consider.

I don't really know, I'm open to org-mode user's opinions like yours in order to make it as natural/compatible as possible.

ericsagnes commented:
FYI I pushed a WIP branch with parsing commands outside of `lib.data` by defining them in a `conf` file that can be overridden: [example](https://github.com/styx-static/styx/blob/rel-1.0.0/src/styx-config.nix).
Still a draft but should be going in the right direction.

Regarding markup, I added dedicated issue: #39

nico202 commented:
The overridable conf is nice! If we solve issue #37 I'll rebase my commits over it

HaoZeke commented:
Hey, sorry, is this still current?

nico202 commented:
Sorry, I moved from nix/nixos to guix, and from styx to haunt. And I've added a draft support for org there too :D
But if you want to superseed me maybe I might still help you

HaoZeke commented:
Well, I'm looking into `styx`, along with `pollen`, so if I do make the switch I'd love to be able to get this PR completed with your help!
